File: 11-container-with-most-water/container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int n = height.size();
        int max_area = 0;
        int left = 0, right = n - 1;

        while (left < right) {
            int h = min(height[left], height[right]);
            int area = h * (right - left);
            max_area = max(max_area, area);
            
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }
        return max_area;
    }
};

File: 11-container-with-most-water/container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        int l = 0;
        int r = height.length - 1;
        int max = 0;
        
        while (l < r) {
            int minH = Math.min(height[l], height[r]);
            int area = minH * (r - l);
            if (area > max) {
                max = area;
            }
            while (l < r && height[l] <= minH) {
                l++;
            }
            while (l < r && height[r] <= minH) {
                r--;
            }
        }
        
        return max;
    }
};